### PR TITLE
Challenge

### PR DESCRIPTION
**Pull request by cito125**
# Made challenge flow for app

- User finds nearby challenges, clicks a challenge which opens a dialog box prompting them to pursue the challenge. Clicking this button opens the camera so they embark on their challenge.

## Upcoming
- Not doing anything with bitmap held in challenges, will wait for firebase to display pictures through the user flow. For now, only hint strings are displayed.
- Right now "nearby challenges" are hardcoded as is the user location. User location will be found using Snapshot API. Nearby challenges will probably remain hardcoded for a bit.
- Implement Fences API logic to search for challenges when user makes fence
- make Challenge activity background turn red when user is out of range of challenge, green when they are within range.
  - Make camera clickable only within 10 meters of challenge (don't know how precise [Haversine formula](https://en.wikipedia.org/wiki/Haversine_formula) is...)

- Once firebase is implemented will send submitted challenges there.

### Note
- Commits involving gradle stuff should probably be squashed or removed since those changes where checked out and push on another branch.

### CODE REVIEW

File: app/src/main/java/com/example/andresarango/aughunt/challenge/challenge_dialog_fragment/ChallengeDialogFragment.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+import com.example.andresarango.aughunt.challenge.Challenge;
+import com.example.andresarango.aughunt.challenge.ChallengeActivity;
+
+public class ChallengeDialogFragment<T> extends DialogFragment {
+
+    private static final String CHALLENGE_HINT = "challenge hint";
+
+    public ChallengeDialogFragment() {
+    }
```
Comment (jrodbx) [inline app/src/main/java/com/example/andresarango/aughunt/challenge/challenge_dialog_fragment/ChallengeDialogFragment.java]: delete.  default constructors are auto-generated when you do not explicitly define one.

File: app/src/main/java/com/example/andresarango/aughunt/challenge/challenge_dialog_fragment/ChallengeDialogFragment.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        });
+
+        alertDialogBuilder.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
+            @Override
+            public void onClick(DialogInterface dialog, int which) {
+                dialog.dismiss();
+            }
+        });
+
+        return alertDialogBuilder.create();
```
Comment (jrodbx) [inline app/src/main/java/com/example/andresarango/aughunt/challenge/challenge_dialog_fragment/ChallengeDialogFragment.java:46]: nit: chain the builder method calls?

File: app/src/main/java/com/example/andresarango/aughunt/challenge/challenges_adapter/ChallengeViewholder.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import android.support.v7.widget.RecyclerView;
+import android.view.View;
+import android.widget.ImageView;
+import android.widget.TextView;
+
+import com.example.andresarango.aughunt.R;
+import com.example.andresarango.aughunt.challenge.Challenge;
+import com.example.andresarango.aughunt.challenge.challenge_dialog_fragment.DialogFragmentListener;
+
+class ChallengeViewholder<T> extends RecyclerView.ViewHolder {
```
Comment (jrodbx) [inline app/src/main/java/com/example/andresarango/aughunt/challenge/challenges_adapter/ChallengeViewholder.java:13]: nit: ViewHolder

File: app/src/main/java/com/example/andresarango/aughunt/challenge/challenges_adapter/ChallengeViewholder.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    public ChallengeViewholder(View itemView) {
+        super(itemView);
+        mHintTextView = (TextView) itemView.findViewById(R.id.viewholder_challenge_hint);
+        mChallengeImageView = (ImageView) itemView.findViewById(R.id.viewholder_challenge_picture);
+    }
+
+    public void bind(Challenge<T> challenge, DialogFragmentListener listener) {
+        mHintTextView.setText(challenge.getHint());
+        mListener = listener;
```
Comment (jrodbx) [inline app/src/main/java/com/example/andresarango/aughunt/challenge/challenges_adapter/ChallengeViewholder.java]: ?

**Review by jrodbx — COMMENTED** [review]
LGTM.

File: app/src/main/java/com/example/andresarango/aughunt/challenge/challenges_adapter/ChallengeViewholder.java (hunk 3 above)
Comment (cito125) [reply app/src/main/java/com/example/andresarango/aughunt/challenge/challenges_adapter/ChallengeViewholder.java:13]: nit?

Comment (jrodbx) [reply app/src/main/java/com/example/andresarango/aughunt/challenge/challenges_adapter/ChallengeViewholder.java:13]: i.e., not important

http://stackoverflow.com/questions/27810522/what-does-nit-mean-in-hacker-speak

File: app/src/main/java/com/example/andresarango/aughunt/challenge/challenge_dialog_fragment/ChallengeDialogFragment.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+import android.content.Intent;
+import android.os.Bundle;
+import android.support.annotation.NonNull;
+import android.support.v4.app.DialogFragment;
+import android.support.v7.app.AlertDialog;
+
+import com.example.andresarango.aughunt.challenge.Challenge;
+import com.example.andresarango.aughunt.challenge.ChallengeActivity;
+
+public class ChallengeDialogFragment<T> extends DialogFragment {
```
Comment (dannylui91) [inline app/src/main/java/com/example/andresarango/aughunt/challenge/challenge_dialog_fragment/ChallengeDialogFragment.java:15]: Not sure why you would need to make this a generic type?

**Review by dannylui91 — APPROVED** [review]
I like it working on my end.

**Pull request merged**